### ISSUE TITLE
Datatable: selectAllRows(), unselectAllRows() partly broken

### ISSUE THREAD

**Issue by Hamsterbau** [primefaces/primefaces#3535]
## 1) Environment
- PrimeFaces version: 6.2.1
- Application server + version: Wildfly 11
- Affected browsers: Firefox (perhaps others)

## 2) Steps to reproduce

1. Click on checkbox in header (calls internally selectAllRows()
2. Unselect one checkbox from one row
3. Nothing happens, as well as pagination and other ajax stuff is broken.
4. After waiting and click another time on row or checkbox it will be unselected

--> More entries means more waiting. It is really slow for lists with > 20k entries.

## 3) Expected behavior

3. Row will be unselected, and objects in selected list will be minus one.

## 4) Sample XHTML

```xhtml
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml"
      xmlns:f="http://xmlns.jcp.org/jsf/core"
      xmlns:p="http://primefaces.org/ui"
      xmlns:h="http://xmlns.jcp.org/jsf/html">

    <h:head />

    <h:body>
        <h:form id="form">
            <p:dataTable
                    id="myTable"
                    widgetVar="myWidget"
                    var="entry"
                    rowKey="#{entry.name}"
                    selection="#{selectableTable.selected}"
                    rows="5"
                    rowSelectMode="add"
                    paginator="true"
                    value="#{selectableTable.entries}">
                <p:ajax event="rowSelect" listener="#{selectableTable.rowSelectCheckbox}" update=":form:selected" />
                <p:ajax event="rowUnselect" listener="#{selectableTable.rowUnselectCheckbox}" update=":form:selected" />
                <p:ajax event="rowSelectCheckbox" listener="#{selectableTable.rowSelectCheckbox}" update=":form:selected" />
                <p:ajax event="rowUnselectCheckbox" listener="#{selectableTable.rowUnselectCheckbox}" update=":form:selected" />
                <p:ajax event="toggleSelect" listener="#{selectableTable.toggleSelected}" update=":form:selected"/>

                <p:column selectionMode="multiple" />

                <p:column sortBy="#{entry.name}">
                    <f:facet name="header">
                        <h:outputText value="Name"/>
                    </f:facet>

                    <h:outputText value="#{entry.name}"/>
                </p:column>

            </p:dataTable>

            <h:outputText id="selected" value="Selected: #{selectableTable.selected.size()}" />
        </h:form>
    </h:body>

</html>
```

## 6) Sample bean

```java
package de.mypackage;

import org.primefaces.context.RequestContext;
import org.primefaces.event.SelectEvent;
import org.primefaces.event.ToggleSelectEvent;
import org.primefaces.event.UnselectEvent;

import javax.annotation.PostConstruct;
import javax.enterprise.context.SessionScoped;
import javax.inject.Named;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

@Named
@SessionScoped
public class SelectableTable implements Serializable{
    private static final long serialVersionUID = 4338079192223741443L;

    private List<Entry> entries;

    private List<Entry> selected;

    public class Entry implements Serializable {
        private static final long serialVersionUID = 910001136636503766L;

        private String name;

        Entry() {}
        Entry(String name) { this.name = name; }

        public String getName() { return name; }
        public void setName(String name) { this.name = name; }
    }

    @PostConstruct
    public void init() {
        this.entries = new ArrayList<>();
        this.entries.add(new Entry("Miriam"));
        this.entries.add(new Entry("Dennis"));
        this.entries.add(new Entry("Dexter"));
        this.entries.add(new Entry("Michelle"));
        this.entries.add(new Entry("Frank"));
        this.entries.add(new Entry("Kurt"));
    }

    public void rowSelectCheckbox(SelectEvent se) {}
    public void rowUnselectCheckbox(UnselectEvent use) {}

    public void toggleSelected(ToggleSelectEvent tse) {
        RequestContext rc = RequestContext.getCurrentInstance();

        if(tse.isSelected()) {
            rc.execute("PF('myWidget').selectAllRows();");
            setSelected(new ArrayList<>(getEntries()));
        } else {
            rc.execute("PF('myWidget').unselectAllRows();");
            setSelected(new ArrayList<>());
        }
    }

    public List<Entry> getEntries() { return entries; }
    public void setEntries(List<Entry> entries) { this.entries = entries; }
    public List<Entry> getSelected() { return selected; }
    public void setSelected(List<Entry> selected) { this.selected = selected; }
}
```

**Comment by tandraschko** [comment]
Does it work without update=@form?
update=@form for such events is actually a nogo.

**Comment by Hamsterbau** [comment]
First, thanks for the quick response. But changing update from **@form** to **:form:selected** after adding id "form" to h:form (to update the selected output field), has the same result.

Should i change the example page to my current test without @form?

**Comment by tandraschko** [comment]
Yep please.

**Comment by melloware** [comment]
@Hamsterbau any update on this issue after you removed updating @form?

**Comment by tandraschko** [comment]
It seems that when deselect a item, the SelectionFeature still receives "@all" instead the "@all - the deselected entry"

**Comment by tandraschko** [comment]
the real problem is the following:
datatable.js removeSelection:
when you deselect a row, it removes the rowkey

but in this case, you selected "all" before and "this.selection" is filled with "@all"
in theory we have to replace "@all" with the rowKeys of all rows and then remove the selected rowKey

it's a simple IF statement actually:

        if (this.selection && this.selection.length === 0 && this.selection[0] === '@all') {
           // do replace
        }

but we don't know all rowKeys as they are loaded lazy....

**Comment by melloware** [comment]
@Hamsterbau I this this workaround is no longer needed if we natively add this feature for `selectionPageOnly="false"` so it selects all items instead of just the page you are on. See #6730

**Comment by melloware** [comment]
OK attached a reproducer for 8.0:
[pf-3535.zip](https://github.com/primefaces/primefaces/files/5752063/pf-3535.zip)

**Comment by melloware** [comment]
@Rapster @stolp OK I have to fix this issue to go with #6730.  @tandraschko analysis of the problem above is correct. 

Here is what happens.

1. First if you check the header checkbox its sets form selection to `form:myTable_selection: @all`.
![image](https://user-images.githubusercontent.com/4399574/103354649-198bf400-4a7a-11eb-9edc-9571ec92c8fc.png)

2. Now uncheck 1 row and it properly unchecks the header box and this value however the value of selection is still `form:myTable_selection: @all` which is wrong. So when you submit this form or use AJAX its going to tell the backend that all rows are still selected.

3. As Thomas pointed out with Lazy we have no idea how many rows there are so we really need to make selection all selected minus the 1 row.   Since there is no way to do that I propose making selection back to all the selected values on the page. So in the below screenshot the selection becomes: `form:myTable_selection: Miriam,Dennis,Michelle,Frank`

![image](https://user-images.githubusercontent.com/4399574/103354780-7a1b3100-4a7a-11eb-9236-0ead3ecec9ce.png)

Its not quite correct but sending `@all` is definitely wrong.  Wanted your thoughts?

**Comment by stolp** [comment]
I can understand the dilemma and cannot think of any better solution. I don't think it will have any influence on most people's usecases, including mine.

Thank you very much for fixing this so quicky! Once it is merged I will verify this with our product.

**Comment by Rapster** [comment]
Giving my 2 cents, but what about whenever `@all` is added to the selection holder:
- unselect a row will add the rowkey to the selection holder
- select a row will remove the rowKey
- to unselect all, uncheck the "select all" checkbox (should clear the selection holder)

On server side, simply do `@all` minus rowKeys? I'm not talking about lazy mode yet, as I don't know how `@all` behave in lazy mode.

**Comment by melloware** [comment]
not quite sure.  Previously eveyrhting just did `selectAllRowsOnPage` the convenience method for `selectAllRows`  was never called before anywhere in JS code it was just provided as a client API widget method and sets '@all`' for the Server Side to interpret.

And unselecting a row tried to find the rowKey in the list and remove it.  But the list just contained `@all` so it removed nothing. And thus the server side still got `@all` which is not correct.

**Comment by tandraschko** [comment]
@melloware reopened as discussed

**Comment by melloware** [comment]
Just for those reading this thread what we discussed is we probably need an `_unselectedRows` sent from the client to the server so we can have...

```
selectRows: '@all',
unselectedRows: '12345'
```

So the server side can determine that all rows should be selected minus the `unselectedRows` and send back to the UI the full list of selections.

**Comment by tandraschko** [comment]
yep exactly
